src/extension/IOController_Posix.hpp

```diff
@@ -35,237 +35,307 @@ namespace extension {
+                        dirty = true;
+                        it    = tasks.erase(it);
+                    }
+                    else {
+                        ++it;
+                    }
+                }
+                else {
+                    ++it;
+                }
+            }
+        }
+
+        /**
+         * @brief Bumps the notification pipe to wake up the poll command
+         *
+         * If the poll command is waiting it will wait forever if something doesn't happen.
+         * When trying to update what to poll or shut down we need to wake it up so it can.
+         */
+        // NOLINTNEXTLINE(readability-make-member-function-const) this changes states
+        void bump() {
+            // Check if there was an error
+            uint8_t val = 1;
+            if (::write(notify_send, &val, sizeof(val)) < 0) {
+                throw std::system_error(network_errno,
+                                        std::system_category(),
+                                        "There was an error while writing to the notification pipe");
+            }
+        }
+
     public:
-        explicit IOController(std::unique_ptr<NUClear::Environment> environment)
-            : Reactor(std::move(environment)), notify_recv(), notify_send() {
+        explicit IOController(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
 
             std::array<int, 2> vals = {-1, -1};
-
-            const int i = ::pipe(vals.data());
+            const int i             = ::pipe(vals.data());
             if (i < 0) {
                 throw std::system_error(network_errno,
                                         std::system_category(),
                                         "We were unable to make the notification pipe for IO");
             }
-
             notify_recv = vals[0];
             notify_send = vals[1];
 
-            // Add our notification pipe to our list of fds
-            fds.push_back(pollfd{notify_recv, POLLIN, 0});
+            // Start by rebuliding the list
+            rebuild_list();
 
             on<Trigger<dsl::word::IOConfiguration>>().then(
                 "Configure IO Reaction",
                 [this](const dsl::word::IOConfiguration& config) {
                     // Lock our mutex to avoid concurrent modification
-                    const std::lock_guard<std::mutex> lock(reaction_mutex);
+                    const std::lock_guard<std::mutex> lock(tasks_mutex);
 
                     // NOLINTNEXTLINE(google-runtime-int)
-                    reactions.emplace_back(config.fd, static_cast<short>(config.events), config.reaction);
+                    tasks.emplace_back(config.fd, static_cast<short>(config.events), config.reaction);
 
                     // Resort our list
-                    std::sort(std::begin(reactions), std::end(reactions));
+                    std::sort(std::begin(tasks), std::end(tasks));
```

Comment:
Change to insertion sort?

Reply:
Should be infrequent enough that it doesn't matter and I'd rather not add an entire insertion sort algorithm rather than just using std::sort